### DISCUSSION TITLE
[RN] Good practice for persistence workflow

### DISCUSSION THREAD

**Discussion by TwistedMinda** [pmndrs/jotai#1166] [Q&A] [answered]
Hello, first of all thanks for this amazing lib, a pleasure to work with.
I'm still learning though, not even discovered a percent of it. I'm wondering here for the correct practice to handle the persistence workflow.

I have used the util `atomWithStorage` with **React Native's AsyncStorage** as per the docs and it's successfully working as expected.
My use-case is pretty simple, i have an action that depends on a stored atom (hence async at first access)

```
const stored = atomWithStorage('my-key', false, ...)
const myAction = atom(null, async (_get, set, payload: any) => {
  const dep = _get(stored)
  await myAsyncAction(_get, set, {dep, payload})
})
```

With no surprise (maybe just a bit), this throws the following error : 
`Reading pending atom state in write operation. We throw a promise for now.`

At this point I thought I would just preload the atoms that I know are required in my actions (we start getting tricky in here)
What I tried is the following: 

```
// Called in top-most component (wrapped by a <Suspense> in the parent)
const list = [stored]
function preload() {
  for (const store of list) {
    useAtomCallback(React.useCallback((get) => {
       return get(store)
    }, []))()
  }
}
```

At least it works but two big caveats:
- I still get the error message because I'm still synchronously reading a pending promise in a "write"-ish operation
- My whole app re-renders each time the "preloaded" atoms change. I was hoping `useAtomCallback` to not trigger a re-render, so I could have just used a simple `useAtomValue` and have the same problem, but at least here we tell the next developer that we didn't intend to use the value in the component itself.

I think I might have missed something, but I'm stuck at this!
Thank you again, much love

**Comment by dai-shi** [comment, marked as answer]
Hi, thanks for opening a discussion.

Per the pending atom error message, do you use async storage? In that case, `delayInit` option may help. (we don't have docs yet. 😓 )

**Reply by TwistedMinda** [reply]
@dai-shi Wow quick response!
I do use async storage indeed, the one information I forgot to mention. I'm on RN!

`delayInit` is a life-saver

```
const storage = { ...createJSONStorage(() => AsyncStorage), delayInit: true }
const loaded = atomWithStorage('my-key', false, storage)
```

Shouldn't it logically be `delayInit: false` because we precisely **don't** want to delay the initialization 🧐 ?

Thanks a lot ! ❤️
Let me know if I can help with the docs.

**Reply by dai-shi** [reply]
> Shouldn't it logically be `delayInit: false` because we precisely **don't** want to delay the initialization 🧐 ?

delayInit=false means it suspends with async storage. This is suspense first. But, it doesn't work in your case because you read it in `write`.
delayInit=true is to disable suspense.
Not quite sure if `delayInit` was the right name, but too late to change it. Let's keep it in the mean time.

> Let me know if I can help with the docs.

Yes, please help us. See #805.

**Reply by TwistedMinda** [reply]
@dai-shi Oh I may i have mistaken something.
And it's actually not working as I expected, no error is thrown and my code runs successfully but the _get() retrieves the default value of the atom at the first run, not the one stored in the AsyncStorage.

EDIT: on a second run of the `write` function, i still don't get a correct value, i receive this:
` {"80224": 4.689135880274525, "80225": 4.572020974981227, "80231": 4.584760732808134, "80253": 4.661946738300763.......`
EDIT2: This was just my fake "Math.random()" that I forgot was populating my atom, lol. 🤦‍♂️

Is my pattern wrong? Or maybe you meant to use `delayInit: true` and still manually preload my atoms as I explained ?
Thanks again

**Reply by dai-shi** [reply]
Please try this:
```js
const stored = atomWithStorage('my-key', false, ...)
const myAction = atom(null, async (get, set, payload: any) => {
  const dep = await get(stored, { unstable_promise: true })
  await myAsyncAction(get, set, {dep, payload})
})
```
This is still an experimental, unstable, undocumented feature to cover such cases.
But, I feel like there should be a better practice for AsyncStorage.

> Or maybe you meant to use `delayInit: true` and still manually preload

Yeah, I would like to see how it goes with `delayInit: true` and manual loading in `write`.

**Reply by TwistedMinda** [reply]
Yep indeed this works.
I proceeded this way:

```
<React.Suspense fallback={null}>
    <Preloader />
    <RestOfApp />
</React.Suspense>
```
with  Preloader being simply:

```
const Preloader = () => {
  useAtomValue(atomToPreload)
  return null
}
```

EDIT: Doesn't feel that bad. Does everything I needed:
- blocks on Suspense with `delayInit:false` (default)
- doesn't block on Suspense with `delayInit:true`

But either way the preloader does the job. 
I just suspect using `delayInit: true` to be very unsafe in case loading from AsyncStorage huge amount of data could be slower than the app workflow that will require the value. 

**> So Suspense looks like the way to me! <**

However would theree be a way to get the value of the atom "just once" to prevent further useless render of the preloader ?

Thank you for your support!

**Reply by dai-shi** [reply]
I think it's a valid practice.
Not sure how good it is. It depends on the suspense practice in general in react 18.
(Yes, it comes with a useless extra re-render.)

Another solution is delayInit: true and manual loading, which is like:
```js
const storage = { ...createJSONStorage(() => AsyncStorage), delayInit: true }
const stored = atomWithStorage('my-key', false, storage)
const myAction = atom(null, async (get, set, payload: any) => {
  let dep = get(stored)
  if (dep === false) {
    dep = await AsyncStorage.getItem('my-key')
  }
  await myAsyncAction(get, set, {dep, payload})
})
```
This comes with an extra `getItem`.

**Reply by TwistedMinda** [reply]
Another cool pattern would be to have an option `autoInit: true` that would simply trigger at `onMount` (we can't even do it manually currently as we only have `set` exposed)
It could play well with `delayInit: true` for people who will refuse to wrap their app in a Suspense.
I think it's the default behaviour with Recoil, as far as I remember using it.

But still same problem for huge amount of data, concurrency may become a thing for some people and that would be an unfair config.
But who ever stores a whole database in his local storage ? Not a minimalist I hope :p 

EDIT: Talking about concurrency... your last reply fixes it with a hacky additional `getItem`. Very smart!
So my problem is fixed. I don't see why not this `autoInit: true` anymore ;)

**Reply by dai-shi** [reply]
> So my problem is fixed.

Great to hear that.

Looking forward to your contribution on docs. 😄
